shared/extension/Behaviour.hpp

```diff
@@ -0,0 +1,317 @@
+     * It should normally be combined with a Causing DSL word.
+     * In this case when another provider needs control and uses a When condition, this will run until that
+     * condition is fulfilled. A Leaving condition will be preferred over an Entering condition that provides the
+     * same Causing statement (the system will rather a module cleans up after itself than ask another module to do
+     * so).
+     *
+     * @tparam T the provider type that this transition function provides for
+     */
+    template <typename T>
+    struct Leaving : public commands::ProviderBase<T, commands::ProviderAction::LEAVING> {};
+
+    /**
+     * This DSL word is used to define a provider for a type.
+     * It will execute in the same way as an on<Trigger<T>> statement except that when it executes will be
+     * determined by the Director. This ensures that it will only run when it has permission to run and will
+     * transition between providers in a sensible way.
+     *
+     * @tparam T the provider type that this transition function provides for
+     */
+    template <typename T>
+    struct Provides : public commands::ProviderBase<T, commands::ProviderAction::NORMAL> {};
+
+    /**
+     * The When DSL word is used to limit access to a provider unless a condition is true.
+     * This will prevent a provider from running if the state is false.
+     * However if there is a Entering<T> or Leaving reaction that has a Causing relationship for this When, it will
+     * be executed to ensure that the Provider can run.
+     *
+     * @tparam T the condition expression that must be true in order to execute this task.
+     */
+    template <typename State, template <typename> class expr, enum State::Value value>
+    struct When {
+
+        template <typename DSL>
+        static inline void bind(const std::shared_ptr<NUClear::threading::Reaction>& reaction) {
+
+            // Tell the director about this when condition
+            reaction->reactor.powerplant.emit(std::make_unique<commands::WhenExpression>(
+                reaction,
+                typeid(State),
+                [](const int& v) { return expr<int>()(v, value); },
+                []() { return NUClear::dsl::operation::CacheGet<State>::get(); },
+                [](NUClear::Reactor& reactor,
+                   std::function<void(const int&)> fn) -> NUClear::threading::ReactionHandle {
+                    return reactor.on<NUClear::dsl::word::Trigger<State>>().then(
+                        [fn](const State& s) { fn(static_cast<int>(s)); });
+                }));
+        }
+    };
+
+    /**
+     * This DSL word is used to create a promise that at some point when running this reaction the state indicated
+     * by T will be true. It may not be on any particular run of the reaction, in which case it will continue to run
+     * that reaction until it is true, or another graph jump becomes necessary.
```

Comment:
Another graph jump? I don't understand that in this context.

Reply:
As in, if A asked for this to run to make some condition true, but then something else happened that rendered the whole operation moot.
E.g. the kick wants the walk to cause a standing state.
But then the robot falls over before that happens.
Nobody cares about the walk getting to a standing state anymore as another graph jump has happened.